processor/resourcedetectionprocessor/internal/aws/eks/detector.go

```diff
@@ -0,0 +1,210 @@
+
+package eks
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/resourcedetectionprocessor/internal"
+)
+
+const (
+	// TypeStr is type of detector.
+	TypeStr = "eks"
+
+	k8sSvcURL = "https://kubernetes.default.svc"
+	/* #nosec */
+	k8sTokenPath      = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	k8sCertPath       = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+	authConfigmapPath = "/api/v1/namespaces/kube-system/configmaps/aws-auth"
+	cwConfigmapPath   = "/api/v1/namespaces/amazon-cloudwatch/configmaps/cluster-info"
+	timeoutMillis     = 2000
+)
+
+// detectorUtils is used for testing the resourceDetector by abstracting functions that rely on external systems.
+type detectorUtils interface {
+	fileExists(filename string) bool
+	fetchString(httpMethod string, URL string) (string, error)
+}
+
+// This struct will implement the detectorUtils interface
+type eksDetectorUtils struct{}
+
+// This struct will help unmarshal clustername from JSON response
+type data struct {
+	ClusterName string `json:"cluster.name"`
+}
+
+var _ internal.Detector = (*Detector)(nil)
+
+// Detector for EKS
+type Detector struct {
+	utils detectorUtils
+}
+
+// Compile time assertion that eksDetectorUtils implements the detectorUtils interface.
+var _ detectorUtils = (*eksDetectorUtils)(nil)
+
+// NewDetector returns a resource detector that will detect AWS EKS resources.
+func NewDetector(_ component.ProcessorCreateParams, _ internal.DetectorConfig) (internal.Detector, error) {
+	return &Detector{utils: &eksDetectorUtils{}}, nil
+}
+
+// Detect returns a Resource describing the Amazon EKS environment being run in.
+func (detector *Detector) Detect(ctx context.Context) (pdata.Resource, error) {
+	res := pdata.NewResource()
+	isEks, err := isEKS(detector.utils)
+	if err != nil {
+		return res, err
+	}
+
+	// Return empty resource object if not running in EKS
+	if !isEks {
+		return res, nil
+	}
+
+	attr := res.Attributes()
+	attr.InsertString(conventions.AttributeCloudProvider, conventions.AttributeCloudProviderAWS)
+	attr.InsertString(conventions.AttributeCloudInfrastructureService, conventions.AttributeCloudProviderAWSEKS)
+
+	// Get clusterName and append to attributes
+	clusterName, err := getClusterName(detector.utils)
+	if err != nil {
+		return res, err
+	}
+	if clusterName != "" {
+		attr.InsertString(conventions.AttributeK8sCluster, clusterName)
+	}
+
+	return res, nil
+}
+
+// isEKS checks if the current environment is running in EKS.
+func isEKS(utils detectorUtils) (bool, error) {
+	if !isK8s(utils) {
+		return false, nil
+	}
+
+	// Make HTTP GET request
+	awsAuth, err := utils.fetchString(http.MethodGet, k8sSvcURL+authConfigmapPath)
```

Comment:
Can we make this call cancellable? We will need to pass a context to this func.
I think it may be important for ensuring Shutdown / Restart of Colletor works correctly.

Reply:
Yes, also needed now that it's going to rely on processor itself cancelling on timeout.